torch_geometric/explainability/explainer.py

```diff
@@ -0,0 +1,221 @@
+
+
+@dataclass
+class Threshold:
+    type: str
+    value: float
+    _valid_type = ["soft", "hard", "connected"]
+
+    def __post_init__(self):
+        if self.type not in self._valid_type:
+            raise ValueError(f'Invalid threshold type {self.type}. '
+                             f'Valid types are {self._valid_type}.')
+        if self.value < 0 or self.value > 1:
+            raise ValueError(f'Invalid threshold value {self.value}. '
+                             f'Valid values are in [0, 1].')
+
+
+@dataclass
+class ExplainerConfig:
+    explanation_type: str
+    mask_type: str
+    _valid_explanation_type = ["model", "phenomenon"]
+    _valid_mask_type = ["node", "edge", "node_and_edge"]
+
+    def __post_init__(self):
+        if self.explanation_type not in self._valid_explanation_type:
+            raise ValueError(
+                f'Invalid explanation type {self.explanation_type}. '
+                f'Valid types are {self._valid_explanation_type}.')
+        if self.mask_type not in self._valid_mask_type:
+            raise ValueError(f'Invalid mask type {self.mask_type}. '
+                             f'Valid types are {self._valid_mask_type}.')
+
+
+class Explainer(torch.nn.Module):
+    r"""A user configuration class for the instance-level explanation
+    of Graph Neural Networks, *e.g.* :class:`~torch_geometric.nn.GCN`.
+
+    Args:
+        explanation_type (str): the type of explanation to be computed.
+            Supported types are `model` and `phenomenon`.
+        explanation_algorithm (ExplainerAlgorithm): the explanation
+            algorithm to be used.
+        mask_type (str): the type of mask to be used. Supported types are
+            `node`, `edge`, `global`. `global` returns masks for both nodes
+            and edges.
+        threshold (str): post-processing thresholding method on the mask.
+            Supported methods are `soft`, `hard`, `connected`.
+        loss (torch.nn.Module): the loss function to be used for the
+            explanation algorithm.
+        model (torch.nn.Module): the GNN module to explain.
+        model_return_type (str): the output type of the model.
+            Supported outputs are `logits`, `probs`, and `regression`.
+        threshold_value (float, optional): the threshold value to be used
+            for the `hard` thresholding method. Should be between 0 and 1.
+            (default: :obj:`0.5`)
+    """
+    def __init__(
+        self,
+        explanation_type: str,
+        # provide a factory instead ?
+        explanation_algorithm: ExplainerAlgorithm,
+        mask_type: str,
+        threshold: str,
+        loss: torch.nn.Module,
+        model: torch.nn.Module,
+        model_return_type: str,
+        threshold_value: float = 0.5,
+    ) -> None:
+
+        super().__init__()
+
+        self.model = model
+        # details of the model
+        if model_return_type.lower() not in ["logits", "probs", "regression"]:
```

Comment:
can multi-label be supported? i.e. a list of probs.

Reply:
We could do it like `captum`, i.e. ask the user for a `target_index`